src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -759,6 +759,8 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
                 throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(SslStream.ReadAsync), "read"));
             }
 
+            Debug.Assert(_internalBuffer is null || _internalBufferCount > 0 || _decryptedBytesCount > 0, "_internalBuffer allocated when no data is buffered.");
```

Comment:
What's the reason for the last part of the condition around `_decryptedBytesCount > 0`?

Reply:
It's because there can be two different kinds of data in the `_internalBuffer` -- encrypted or decrypted -- and the two checks here test for each.

The field names aren't ideal; `_internalBufferCount` should probably be something like `_encryptedBytesCount`.

Note that `ReturnReadBufferIfEmpty` (just above) does the same check before returning the unused buffer to the pool.